Add Dependency Extraction Webpack Plugin

psealock opened the pull request:
Fixes https://github.com/woocommerce/woocommerce-admin/issues/5647

Changes our script dependencies to use our own Dependency Extraction Webpack Plugin. This PR updates the main script, packages, and wc-admin scripts.

### Detailed test instructions:

1. `npm start`
1. Make sure the app builds and no fatal errors occur as you visit screens and pages.
2. Check the console for errors relating to js files not being found.
3. Visit the coupons page to ensure that wc-admin-scripts are also working.

### Bonus Testing

1. Add a new dependency to a package. ie, add `import { compose } from '@wordpress/compose';` to any package that would have that dependency, such as `@woocommerce/currency`. Note, you may need to log it out so it gets used, otherwise webpack may disregard it.
2. Confirm the presence of `wp-compose` in `dist/currency/index.min.assets.php`

### Changelog Note:

- Dev: Use Woo Dependency Extraction Webpack Plugin to gather dependencies

joshuatf reviewed (commented):
Awesome work, Paul!  This is testing really well for me.

One more addition that would be really great to have around this is some documentation on how this works or how to implement, but otherwise I ran into no issues in smoke testing.

psealock commented:
@joshuatf thanks for the review. There is documentation about the package in https://github.com/woocommerce/woocommerce-admin/blob/main/packages/dependency-extraction-webpack-plugin/README.md.

That readme notes the additional features of the Woo flavor while pointing to Gutenberg's readme of the same package for more information about the api. Do you think that suffices or is more required in wc-admin?

joshuatf reviewed (commented):
I think the link to the original Dependency Extraction Webpack Plugin docs is enough.  

After rebasing on `main` I'm getting this error and WCA fails to load.  Can you confirm that you see this or is this an issue with my local env?

psealock commented:
I rebased against `main` and did not see an issue. Once you rebase, you may need to re run `npm start` as a dependency may need to get recalculated?

joshuatf reviewed (commented):
Still getting the same error when rebasing from the latest `main`.  This branch works as is, but after rebasing I'm not able to see anything.

<img width="530" alt="Screen Shot 2021-01-07 at 2 30 47 PM" src="https://user-images.githubusercontent.com/10561050/103969659-659efe80-5134-11eb-9bf7-39f7368cb7c3.png">

joshuatf commented:
Maybe someone else on @woocommerce/isotope can rebase on the latest `main` and see if they get the same error or if there's something wrong with my environment.

joelclimbsthings commented:
If so, it's wrong on mine as well @joshuatf . Looks like the same error after rebasing on main: 

![image](https://user-images.githubusercontent.com/444632/104049387-c1e83980-5199-11eb-858a-92a9267bca0b.png)

psealock commented:
I rebased again and could not reproduce the error. I built a zip and tested on JN and could also not reproduce the error.

Incidentally, I discovered that asset files weren't included in builds, so I added them in https://github.com/woocommerce/woocommerce-admin/pull/5762/commits/fd99ead78723aa0f3e8e43c8bd9dbe123dee1098.

I'm wondering if you have a plugin with incorrect dependencies declared, potentially one of the test ones used in previous PRs that add extensibility to the Nav. They may have been working correctly previously only due to coincidence. Could you disable plugins and see if that resolves the issue?

joshuatf commented:
>I'm wondering if you have a plugin with incorrect dependencies declared, potentially one of the test ones used in previous PRs that add extensibility to the Nav. 

Good call.  I disabled all plugins and no changes, however, I was running on WC 4.8.0, upgrading to WC 4.9.0 resolved the issue.

Is there a way to continue supporting WC 4.8.0 if this is the root problem?

psealock commented:
>Is there a way to continue supporting WC 4.8.0 if this is the root problem?

Thanks for figuring this out Josh. Between this issue and the bug that I found and fixed yesterday I think its best to hold off on this change until after the 1.9.x branch has been cut. It's too far reaching of a change to be added in this late in the cycle.

>Is there a way to continue supporting WC 4.8.0 if this is the root problem?

Interesting that it didn't break until a rebase was involved. At least there are steps to reproduce that I will look into.

psealock commented:
@joshuatf @joelclimbsthings, I rebased yet again and tested with WC 4.8 and even with WP 5.4 and can't reproduce the error.
I also tried these combinations on JN with no luck either. The end to end tests are failing with the same error messages though, so trying to figure out what the issue could be.

psealock commented:
I'm marking this as ready for review. I still can't reproduce any errors and now CI tests are all passing.

samueljseay reviewed (approved):
tests well! Just one suggestion, not a blocker though. 🚀

psealock commented on src/Loader.php:
```diff
@@ -301,120 +301,44 @@ public static function register_scripts() {
+				$script_assets ['dependencies'],
 				$js_file_version,
 				true
 			);
```
Good idea, added in https://github.com/woocommerce/woocommerce-admin/pull/5762/commits/73443699de8eca3d50693796366cafdd4e84263e
